File: Stripe/STPAddressViewModel.m
//
//  STPAddressViewModel.m
//  Stripe
//
//  Created by Jack Flintermann on 4/21/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

#import "STPAddressViewModel.h"

#import "NSArray+Stripe.h"
#import "STPDispatchFunctions.h"
#import "STPPostalCodeValidator.h"

#import <CoreLocation/CoreLocation.h>

@interface STPAddressViewModel()<STPAddressFieldTableViewCellDelegate>
@property (nonatomic) BOOL isBillingAddress;
@property (nonatomic) STPBillingAddressFields requiredBillingAddressFields;
@property (nonatomic) NSSet<STPContactField> *requiredShippingAddressFields;
@property (nonatomic) NSArray<STPAddressFieldTableViewCell *> *addressCells;
@property (nonatomic) BOOL showingPostalCodeCell;
@property (nonatomic) BOOL geocodeInProgress;
@end

@implementation STPAddressViewModel

@synthesize addressFieldTableViewCountryCode = _addressFieldTableViewCountryCode;
@synthesize availableCountries = _availableCountries;

- (instancetype)initWithRequiredBillingFields:(STPBillingAddressFields)requiredBillingAddressFields availableCountries:(NSSet<NSString *> *)availableCountries {
    self = [super init];
    if (self) {
        _isBillingAddress = YES;
        _availableCountries = [availableCountries copy];
        _requiredBillingAddressFields = requiredBillingAddressFields;
        switch (requiredBillingAddressFields) {
            case STPBillingAddressFieldsNone:
                _addressCells = @[];
                break;
            case STPBillingAddressFieldsPostalCode:
                _addressCells = @[
                                  // Postal code cell will be added later if necessary
                                  ];
                break;
            case STPBillingAddressFieldsFull:
                _addressCells = @[
                                  [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeName contents:@"" lastInList:NO delegate:self],
                                  [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeLine1 contents:@"" lastInList:NO delegate:self],
                                  [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeLine2 contents:@"" lastInList:NO delegate:self],
                                  [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeCountry contents:_addressFieldTableViewCountryCode lastInList:NO delegate:self],
                                  // Postal code cell will be added here later if necessary
                                  [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeCity contents:@"" lastInList:NO delegate:self],
                                  [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeState contents:@"" lastInList:YES delegate:self],
                                  ];
                break;
            case STPBillingAddressFieldsName:
                _addressCells = @[
                                  [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeName contents:@"" lastInList:YES delegate:self]
                                  ];
                break;
        }
        [self commonInit];
    }
    return self;
}

- (instancetype)initWithRequiredShippingFields:(NSSet<STPContactField> *)requiredShippingAddressFields availableCountries:(NSSet<NSString *> *)availableCountries {
    self = [super init];
    if (self) {
        _isBillingAddress = NO;
        _availableCountries = [availableCountries copy];
        _requiredShippingAddressFields = requiredShippingAddressFields;
        NSMutableArray *cells = [NSMutableArray new];
        if ([requiredShippingAddressFields containsObject:STPContactFieldName]) {
            [cells addObject:[[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeName contents:@"" lastInList:NO delegate:self]];
        }
        if ([requiredShippingAddressFields containsObject:STPContactFieldEmailAddress]) {
            [cells addObject:[[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeEmail contents:@"" lastInList:NO delegate:self]];
        }
        if ([requiredShippingAddressFields containsObject:STPContactFieldPostalAddress]) {
            NSMutableArray *postalCells = [@[
                                             [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeName contents:@"" lastInList:NO delegate:self],
                                             [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeLine1 contents:@"" lastInList:NO delegate:self],
                                             [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeLine2 contents:@"" lastInList:NO delegate:self],
                                             [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeCountry contents:_addressFieldTableViewCountryCode lastInList:NO delegate:self],
                                             // Postal code cell will be added here later if necessary
                                             [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeCity contents:@"" lastInList:NO delegate:self],
                                             [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeState contents:@"" lastInList:NO delegate:self],
                                             ] mutableCopy];
            if ([requiredShippingAddressFields containsObject:STPContactFieldName]) {
                [postalCells removeObjectAtIndex:0];
            }
            [cells addObjectsFromArray:postalCells];
        }
        if ([requiredShippingAddressFields containsObject:STPContactFieldPhoneNumber]) {
            [cells addObject:[[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypePhone contents:@"" lastInList:NO delegate:self]];
        }
        STPAddressFieldTableViewCell *lastCell = [cells lastObject];
        if (lastCell != nil) {
            lastCell.lastInList = YES;
        }
        _addressCells = [cells copy];
        [self commonInit];
    }
    return self;
}

- (instancetype)initWithRequiredBillingFields:(STPBillingAddressFields)requiredBillingAddressFields {
    return [self initWithRequiredBillingFields:requiredBillingAddressFields availableCountries:nil];
}
- (instancetype)initWithRequiredShippingFields:(NSSet<STPContactField> *)requiredShippingAddressFields {
    return [self initWithRequiredShippingFields:requiredShippingAddressFields availableCountries:nil];
}

- (void)commonInit {
    _addressFieldTableViewCountryCode = [[NSLocale autoupdatingCurrentLocale] objectForKey:NSLocaleCountryCode];
    [self updatePostalCodeCellIfNecessary];
}

- (void)updatePostalCodeCellIfNecessary {
    [self.delegate addressViewModelWillUpdate:self];
    BOOL shouldBeShowingPostalCode = [STPPostalCodeValidator postalCodeIsRequiredForCountryCode:self.addressFieldTableViewCountryCode];
    if (shouldBeShowingPostalCode && !self.showingPostalCodeCell) {
        if (self.containsStateAndPostalFields) {
            // Add before city
            NSUInteger stateFieldIndex = [self.addressCells indexOfObjectPassingTest:^BOOL(STPAddressFieldTableViewCell * _Nonnull obj, NSUInteger __unused idx, BOOL * _Nonnull __unused stop) {
                return (obj.type == STPAddressFieldTypeCity);
            }];

            if (stateFieldIndex != NSNotFound) {
                NSUInteger zipFieldIndex = stateFieldIndex;

                NSMutableArray<STPAddressFieldTableViewCell *> *mutableAddressCells = self.addressCells.mutableCopy;
                [mutableAddressCells insertObject:[[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeZip contents:@"" lastInList:NO delegate:self]
                                          atIndex:zipFieldIndex];
                self.addressCells = mutableAddressCells.copy;
                [self.delegate addressViewModel:self addedCellAtIndex:zipFieldIndex];
                [self.delegate addressViewModelDidChange:self];
            }
        }
    } else if (!shouldBeShowingPostalCode && self.showingPostalCodeCell) {
        if (self.containsStateAndPostalFields) {
            NSUInteger zipFieldIndex = [self.addressCells indexOfObjectPassingTest:^BOOL(STPAddressFieldTableViewCell * _Nonnull obj, NSUInteger __unused idx, BOOL * _Nonnull __unused stop) {
                return (obj.type == STPAddressFieldTypeZip);
            }];

            if (zipFieldIndex != NSNotFound) {
                NSMutableArray<STPAddressFieldTableViewCell *> *mutableAddressCells = self.addressCells.mutableCopy;
                [mutableAddressCells removeObjectAtIndex:zipFieldIndex];
                self.addressCells = mutableAddressCells.copy;
                [self.delegate addressViewModel:self removedCellAtIndex:zipFieldIndex];
                [self.delegate addressViewModelDidChange:self];
            }
        }
    }
    self.showingPostalCodeCell = shouldBeShowingPostalCode;
    [self.delegate addressViewModelDidUpdate:self];
}

- (BOOL)containsStateAndPostalFields {
    if (self.isBillingAddress) {
        return self.requiredBillingAddressFields == STPBillingAddressFieldsFull;
    } else {
        return [self.requiredShippingAddressFields containsObject:STPContactFieldPostalAddress];
    }
}

- (STPAddressFieldTableViewCell *)cellAtIndex:(NSInteger)index {
    return self.addressCells[index];
}

- (void)addressFieldTableViewCellDidReturn:(STPAddressFieldTableViewCell *)cell {
    [[self cellAfterCell:cell] becomeFirstResponder];
}

- (void)addressFieldTableViewCellDidEndEditing:(STPAddressFieldTableViewCell *)cell {
    if (cell.type == STPAddressFieldTypeZip) {
        [self updateCityAndStateFromZipCodeCell:cell];
    }
}

- (void)updateCityAndStateFromZipCodeCell:(STPAddressFieldTableViewCell *)zipCell {

    NSString *zipCode = zipCell.contents;

    if (self.geocodeInProgress
        || zipCode == nil
        || !zipCell.textField.validText
        || ![_addressFieldTableViewCountryCode isEqualToString:@"US"]) {
        return;
    }

    STPAddressFieldTableViewCell *cityCell = nil;
    STPAddressFieldTableViewCell *stateCell = nil;
    for (STPAddressFieldTableViewCell *cell in self.addressCells) {
        if (cell.type == STPAddressFieldTypeCity) {
            cityCell = cell;
        } else if (cell.type == STPAddressFieldTypeState) {
            stateCell = cell;
        }
    }

    if ((cityCell == nil && stateCell == nil)
        || (cityCell.contents.length > 0 || stateCell.contents.length > 0)) {
        // Don't auto fill if either have text already
        // Or if neither are non-nil
        return;
    } else {
        self.geocodeInProgress = YES;
        CLGeocoder *geocoder = [CLGeocoder new];

        CLGeocodeCompletionHandler onCompletion = ^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error) {
            stpDispatchToMainThreadIfNecessary(^{
                if (placemarks.count > 0 && error == nil) {
                    CLPlacemark *placemark = placemarks.firstObject;
                    if (cityCell.contents.length == 0
                        && stateCell.contents.length == 0
                        && [zipCell.contents isEqualToString:zipCode]) {
                        // Check contents again to make sure they're still empty
                        // And that zipcode hasn't changed to something else
                        cityCell.contents = placemark.locality;
                        stateCell.contents = placemark.administrativeArea;
                    }
                }
                self.geocodeInProgress = NO;
            });
        };

        if (@available(iOS 11, *)) {
            CNMutablePostalAddress *address = [CNMutablePostalAddress new];
            address.postalCode = zipCode;
            address.ISOCountryCode = _addressFieldTableViewCountryCode;

            [geocoder geocodePostalAddress:address.copy
                         completionHandler:onCompletion];
        } else {
            [geocoder geocodeAddressString:[NSString stringWithFormat:@"%@, %@", zipCode, _addressFieldTableViewCountryCode]
                         completionHandler:onCompletion];
        }
    }
}

- (void)addressFieldTableViewCellDidUpdateText:(__unused STPAddressFieldTableViewCell *)cell {
    [self.delegate addressViewModelDidChange:self];
}

- (BOOL)isValid {
    if (self.isBillingAddress) {
        if (self.requiredBillingAddressFields == STPBillingAddressFieldsPostalCode) {
            return YES; // The AddressViewModel is only for address fields. Determining whether the postal code is present is up to the STPCardTextFieldViewModel.
        }
        return [self.address containsRequiredFields:self.requiredBillingAddressFields];
    } else {
        return [self.address containsRequiredShippingAddressFields:self.requiredShippingAddressFields];
    }
}

- (void)setAddressFieldTableViewCountryCode:(NSString *)addressFieldTableViewCountryCode {
    if (addressFieldTableViewCountryCode.length > 0 // ignore if someone passing in nil or empty and keep our current setup
        && ![_addressFieldTableViewCountryCode isEqualToString:addressFieldTableViewCountryCode]) {
        _addressFieldTableViewCountryCode = addressFieldTableViewCountryCode.copy;
        [self updatePostalCodeCellIfNecessary];
        for (STPAddressFieldTableViewCell *cell in self.addressCells) {
            [cell delegateCountryCodeDidChange:_addressFieldTableViewCountryCode];
        }
    }
}

- (void)setAddress:(STPAddress *)address {
    self.addressFieldTableViewCountryCode = address.country;

    for (STPAddressFieldTableViewCell *cell in self.addressCells) {
        switch (cell.type) {
            case STPAddressFieldTypeName:
                cell.contents = address.name;
                break;
            case STPAddressFieldTypeLine1:
                cell.contents = address.line1;
                break;
            case STPAddressFieldTypeLine2:
                cell.contents = address.line2;
                break;
            case STPAddressFieldTypeCity:
                cell.contents = address.city;
                break;
            case STPAddressFieldTypeState:
                cell.contents = address.state;
                break;
            case STPAddressFieldTypeZip:
                cell.contents = address.postalCode;
                break;
            case STPAddressFieldTypeCountry:
                cell.contents = address.country;
                break;
            case STPAddressFieldTypeEmail:
                cell.contents = address.email;
                break;
            case STPAddressFieldTypePhone:
                cell.contents = address.phone;
                break;
        }
    }
}

- (STPAddress *)address {
    STPAddress *address = [STPAddress new];
    for (STPAddressFieldTableViewCell *cell in self.addressCells) {

        switch (cell.type) {
            case STPAddressFieldTypeName:
                address.name = cell.contents;
                break;
            case STPAddressFieldTypeLine1:
                address.line1 = cell.contents;
                break;
            case STPAddressFieldTypeLine2:
                address.line2 = cell.contents;
                break;
            case STPAddressFieldTypeCity:
                address.city = cell.contents;
                break;
            case STPAddressFieldTypeState:
                address.state = cell.contents;
                break;
            case STPAddressFieldTypeZip:
                address.postalCode = cell.contents;
                break;
            case STPAddressFieldTypeCountry:
                address.country = cell.contents;
                break;
            case STPAddressFieldTypeEmail:
                address.email = cell.contents;
                break;
            case STPAddressFieldTypePhone:
                address.phone = cell.contents;
                break;
        }
    }
    // Prefer to use the contents of STPAddressFieldTypeCountry, but fallback to
    // `addressFieldTableViewCountryCode` if nil (important for STPBillingAddressFieldsPostalCode)
    address.country = address.country ?: self.addressFieldTableViewCountryCode;
    return address;
}

- (STPAddressFieldTableViewCell *)cellAfterCell:(STPAddressFieldTableViewCell *)cell {
    NSInteger index = [self.addressCells indexOfObject:cell];
    return [self.addressCells stp_boundSafeObjectAtIndex:index + 1];
}

@end

File: Stripe/STPAddressViewModel.swift
//
//  STPAddressViewModel.swift
//  Stripe
//
//  Created by Jack Flintermann on 4/21/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

import Contacts
import CoreLocation
import UIKit

protocol STPAddressViewModelDelegate: class {
  func addressViewModelDidChange(_ addressViewModel: STPAddressViewModel)
  func addressViewModel(_ addressViewModel: STPAddressViewModel, addedCellAt index: Int)
  func addressViewModel(_ addressViewModel: STPAddressViewModel, removedCellAt index: Int)
  func addressViewModelWillUpdate(_ addressViewModel: STPAddressViewModel)
  func addressViewModelDidUpdate(_ addressViewModel: STPAddressViewModel)
}

class STPAddressViewModel: STPAddressFieldTableViewCellDelegate {
  private(set) var addressCells: [STPAddressFieldTableViewCell] = []
  weak var delegate: STPAddressViewModelDelegate?

  var addressFieldTableViewCountryCode: String? = Locale.autoupdatingCurrent.regionCode ?? "" {
    didSet {
      updatePostalCodeCellIfNecessary()
      if let addressFieldTableViewCountryCode = addressFieldTableViewCountryCode {
        for cell in addressCells {
          cell.delegateCountryCodeDidChange(countryCode: addressFieldTableViewCountryCode)
        }
      }
    }
  }

  var address: STPAddress {
    get {
      let address = STPAddress()
      for cell in addressCells {

        switch cell.type {
        case .name:
          address.name = cell.contents
        case .line1:
          address.line1 = cell.contents
        case .line2:
          address.line2 = cell.contents
        case .city:
          address.city = cell.contents
        case .state:
          address.state = cell.contents
        case .zip:
          address.postalCode = cell.contents
        case .country:
          address.country = cell.contents
        case .email:
          address.email = cell.contents
        case .phone:
          address.phone = cell.contents
        }
      }
      // Prefer to use the contents of STPAddressFieldTypeCountry, but fallback to
      // `addressFieldTableViewCountryCode` if nil (important for STPBillingAddressFieldsPostalCode)
      address.country = address.country ?? addressFieldTableViewCountryCode
      return address
    }
    set(address) {
      if let country = address.country {
        addressFieldTableViewCountryCode = country
      }

      for cell in addressCells {
        switch cell.type {
        case .name:
          cell.contents = address.name
        case .line1:
          cell.contents = address.line1
        case .line2:
          cell.contents = address.line2
        case .city:
          cell.contents = address.city
        case .state:
          cell.contents = address.state
        case .zip:
          cell.contents = address.postalCode
        case .country:
          cell.contents = address.country
        case .email:
          cell.contents = address.email
        case .phone:
          cell.contents = address.phone
        }
      }
    }
  }

  /* The default value of availableCountries is nil, which will allow all known countries. */
  var availableCountries: Set<String>?

  var isValid: Bool {
    if isBillingAddress {
      if requiredBillingAddressFields == .postalCode {
        return true  // The AddressViewModel is only for address fields. Determining whether the postal code is present is up to the STPCardTextFieldViewModel.
      } else {
        return address.containsRequiredFields(requiredBillingAddressFields)
      }

    } else {
      if let requiredShippingAddressFields = requiredShippingAddressFields {
        return address.containsRequiredShippingAddressFields(requiredShippingAddressFields)
      }
      return false
    }
  }

  // The default value of availableCountries is nil, which will allow all known countries.
  init(
    requiredBillingFields requiredBillingAddressFields: STPBillingAddressFields,
    availableCountries: Set<String>? = nil
  ) {
    isBillingAddress = true
    self.availableCountries = availableCountries
    self.requiredBillingAddressFields = requiredBillingAddressFields
    switch requiredBillingAddressFields {
    case .none:
      addressCells = []
    case .zip, .postalCode:
      addressCells = []  // Postal code cell will be added later if necessary
    case .full:
      addressCells = [
        STPAddressFieldTableViewCell(type: .name, contents: "", lastInList: false, delegate: self),
        STPAddressFieldTableViewCell(type: .line1, contents: "", lastInList: false, delegate: self),
        STPAddressFieldTableViewCell(type: .line2, contents: "", lastInList: false, delegate: self),
        STPAddressFieldTableViewCell(
          type: .country, contents: addressFieldTableViewCountryCode, lastInList: false,
          delegate: self),
        // Postal code cell will be added here later if necessary
        STPAddressFieldTableViewCell(type: .city, contents: "", lastInList: false, delegate: self),
        STPAddressFieldTableViewCell(type: .state, contents: "", lastInList: true, delegate: self),
      ]
    case .name:
      addressCells = [
        STPAddressFieldTableViewCell(type: .name, contents: "", lastInList: true, delegate: self)
      ]
    default:
      fatalError()
    }
    commonInit()
  }

  init(
    requiredShippingFields requiredShippingAddressFields: Set<STPContactField>,
    availableCountries: Set<String>? = nil
  ) {
    isBillingAddress = false
    self.availableCountries = availableCountries
    self.requiredShippingAddressFields = requiredShippingAddressFields
    var cells: [STPAddressFieldTableViewCell] = []

    if requiredShippingAddressFields.contains(STPContactField.name) {
      cells.append(
        STPAddressFieldTableViewCell(type: .name, contents: "", lastInList: false, delegate: self))
    }
    if requiredShippingAddressFields.contains(.emailAddress) {
      cells.append(
        STPAddressFieldTableViewCell(type: .email, contents: "", lastInList: false, delegate: self))
    }
    if requiredShippingAddressFields.contains(STPContactField.postalAddress) {
      var postalCells = [
        STPAddressFieldTableViewCell(type: .name, contents: "", lastInList: false, delegate: self),
        STPAddressFieldTableViewCell(type: .line1, contents: "", lastInList: false, delegate: self),
        STPAddressFieldTableViewCell(type: .line2, contents: "", lastInList: false, delegate: self),
        STPAddressFieldTableViewCell(
          type: .country, contents: addressFieldTableViewCountryCode, lastInList: false,
          delegate: self),
        // Postal code cell will be added here later if necessary
        STPAddressFieldTableViewCell(type: .city, contents: "", lastInList: false, delegate: self),
        STPAddressFieldTableViewCell(type: .state, contents: "", lastInList: false, delegate: self),
      ]
      if requiredShippingAddressFields.contains(.name) {
        postalCells.remove(at: 0)
      }
      cells.append(contentsOf: postalCells.compactMap { $0 })
    }
    if requiredShippingAddressFields.contains(.phoneNumber) {
      cells.append(
        STPAddressFieldTableViewCell(type: .phone, contents: "", lastInList: false, delegate: self))
    }
    if let lastCell = cells.last {
      lastCell.lastInList = true
    }
    addressCells = cells
    commonInit()
  }

  private func cell(at index: Int) -> STPAddressFieldTableViewCell? {
    guard index > 0,
      index < addressCells.count
    else {
      return nil
    }
    return addressCells[index]
  }

  private var isBillingAddress = false
  private var requiredBillingAddressFields: STPBillingAddressFields = .none
  private var requiredShippingAddressFields: Set<STPContactField>?
  private var showingPostalCodeCell = false
  private var geocodeInProgress = false

  private func commonInit() {
    if let countryCode = Locale.autoupdatingCurrent.regionCode {
      addressFieldTableViewCountryCode = countryCode
    } else {
      addressFieldTableViewCountryCode = ""
    }
    updatePostalCodeCellIfNecessary()
  }

  private func updatePostalCodeCellIfNecessary() {
    delegate?.addressViewModelWillUpdate(self)
    let shouldBeShowingPostalCode = STPPostalCodeValidator.postalCodeIsRequired(
      forCountryCode: addressFieldTableViewCountryCode)

    if shouldBeShowingPostalCode && !showingPostalCodeCell {
      if containsStateAndPostalFields() {
        // Add before city
        let zipFieldIndex = addressCells.firstIndex(where: { $0.type == .city }) ?? 0

        var mutableAddressCells = addressCells
        mutableAddressCells.insert(
          STPAddressFieldTableViewCell(type: .zip, contents: "", lastInList: false, delegate: self),
          at: zipFieldIndex)
        addressCells = mutableAddressCells
        delegate?.addressViewModel(self, addedCellAt: zipFieldIndex)
        delegate?.addressViewModelDidChange(self)
      }
    } else if !shouldBeShowingPostalCode && showingPostalCodeCell {
      if containsStateAndPostalFields() {
        if let zipFieldIndex = addressCells.firstIndex(where: { $0.type == .zip }) {

          var mutableAddressCells = addressCells
          mutableAddressCells.remove(at: zipFieldIndex)
          addressCells = mutableAddressCells
          delegate?.addressViewModel(self, removedCellAt: zipFieldIndex)
          delegate?.addressViewModelDidChange(self)
        }
      }
    }
    showingPostalCodeCell = shouldBeShowingPostalCode
    delegate?.addressViewModelDidUpdate(self)
  }

  private func containsStateAndPostalFields() -> Bool {
    if isBillingAddress {
      return requiredBillingAddressFields == .full
    } else {
      return requiredShippingAddressFields?.contains(.postalAddress) ?? false
    }
  }

  func updateCityAndState(fromZipCodeCell zipCell: STPAddressFieldTableViewCell?) {

    let zipCode = zipCell?.contents

    if geocodeInProgress || zipCode == nil || !(zipCell?.textField?.validText ?? false)
      || !(addressFieldTableViewCountryCode == "US")
    {
      return
    }

    var cityCell: STPAddressFieldTableViewCell?
    var stateCell: STPAddressFieldTableViewCell?
    for cell in addressCells {
      if cell.type == .city {
        cityCell = cell
      } else if cell.type == .state {
        stateCell = cell
      }
    }

    if (cityCell == nil && stateCell == nil)
      || ((cityCell?.contents?.count ?? 0) > 0 || (stateCell?.contents?.count ?? 0) > 0)
    {
      // Don't auto fill if either have text already
      // Or if neither are non-nil
      return
    } else {
      geocodeInProgress = true
      let geocoder = CLGeocoder()

      let onCompletion: CLGeocodeCompletionHandler = { placemarks, error in
        stpDispatchToMainThreadIfNecessary({
          if (placemarks?.count ?? 0) > 0 && error == nil {
            let placemark = placemarks?.first
            if (cityCell?.contents?.count ?? 0) == 0 && (stateCell?.contents?.count ?? 0) == 0
              && (zipCell?.contents == zipCode)
            {
              // Check contents again to make sure they're still empty
              // And that zipcode hasn't changed to something else
              cityCell?.contents = placemark?.locality
              stateCell?.contents = placemark?.administrativeArea
            }
          }
          self.geocodeInProgress = false
        })
      }

      let address = CNMutablePostalAddress()
      address.postalCode = zipCode ?? ""
      address.isoCountryCode = addressFieldTableViewCountryCode ?? ""

      geocoder.geocodePostalAddress(
        address,
        completionHandler: onCompletion)
    }
  }

  private func cell(after cell: STPAddressFieldTableViewCell?) -> STPAddressFieldTableViewCell? {
    guard let cell = cell,
      let cellIndex = addressCells.firstIndex(of: cell),
      cellIndex + 1 < addressCells.count
    else {
      return nil
    }
    return addressCells[cellIndex + 1]
  }

  func addressFieldTableViewCellDidUpdateText(_ cell: STPAddressFieldTableViewCell?) {
    delegate?.addressViewModelDidChange(self)
  }

  func addressFieldTableViewCellDidReturn(_ cell: STPAddressFieldTableViewCell?) {
    _ = self.cell(after: cell)?.becomeFirstResponder()
  }

  func addressFieldTableViewCellDidEndEditing(_ cell: STPAddressFieldTableViewCell?) {
    if cell?.type == .zip {
      updateCityAndState(fromZipCodeCell: cell)
    }
  }

}
